src/react/databrowser/objects/ObjectUpload.jsx

```diff
@@ -0,0 +1,157 @@
+import { formatBytes } from '../../utils';
+import styled from 'styled-components';
+import { useDropzone } from 'react-dropzone';
+
+const DropZone = styled.div`
+    flex: 1;
+    display: flex;
+    flex-direction: column;
+
+    height: 300px;
+    width: 500px;
+    padding: 20px;
+    border-width: 2px;
+    border-radius: 2px;
+    border-color: ${props => props.theme.brand.borderLight};
+    border-style: dashed;
+`;
+
+const RemoveCell = styled(T.Cell)`
+    width: 10px;
+`;
+
+const Files = styled.div`
+    height: 250px;
+    overflow-y: scroll;
+    margin: 10px 0px;
+`;
+
+const EmptyFile = styled.div`
+    text-align: center;
+    margin-top: 60px;
+    & > * {
+        margin-bottom: 15px;
+    }
+`;
+
+const title = size => !size ? 'Upload' : `Upload ${size} file${size > 1 ? 's': ''}`;
+
+type FileListProps = {
+    acceptedFiles: Array<File>,
+    fileRejections: Array<File>,
+    open: () => void,
+    removeFile: (string) => void,
+};
+// NOTE: If accept, multiple, minSize or maxSize are added to useDropzone parameters,
+// files might be rejected (fileRejections) and we will have to show them in <FileList/>.
+const FileList = ({ acceptedFiles, open, removeFile }: FileListProps) =>
+    <div>
+        <Button icon={<i className="fas fa-plus" />} size='small' text='Add more files' variant='info' onClick={open} />
+        <Files>
+            <Table>
+                <T.Body>
+                    {
+                        acceptedFiles.map(file => (
+                            <T.Row key={file.path}>
+                                <T.Cell> {file.path} <br/> <small>{formatBytes(file.size)}</small> </T.Cell>
+                                <RemoveCell> <div onClick={() => removeFile(file.path)}> <i className='fas fa-times'></i> </div> </RemoveCell>
+                            </T.Row>
+                        ))
+                    }
+                </T.Body>
+            </Table>
+        </Files>
+    </div>;
+
+type NoFileProps = {
+    open: () => void,
+};
+const NoFile = ({ open }: NoFileProps) =>
+    <EmptyFile>
+        <i className="fas fa-3x fa-file-upload"></i>
+        <div> Drag and drop files and folders here </div>
+        <div> OR </div>
+        <Button icon={<i className="fas fa-plus" />} text='Add files' variant='info' onClick={open} />
+    </EmptyFile>;
+
+type Props = {
+    bucketName: string,
+    prefixWithSlash: string,
+};
+const ObjectUpload = ({ bucketName, prefixWithSlash }: Props) => {
+    const [acceptedFiles, setAcceptedFiles] = useState([]);
+    const [fileRejections, setFileRejections] = useState([]);
+    const show = useSelector((state: AppState) => state.uiObjects.showObjectUpload);
+    const dispatch: DispatchAPI<Action> = useDispatch();
+
+    const onDrop = (accepted, rejections) => {
+        if (accepted.length > 0) {
+            setAcceptedFiles([...acceptedFiles, ...accepted.filter(a => !acceptedFiles.find(f => f.path === a.path))]);
```

Comment:
What if you filtered before setting the new files? Is there a chance that the filter clears all items in the array making no change? Like, even if you did drop some accepted items if they were already in the list this would do nothing? I guess it does nothing each time but if I remember correctly react will handle a few things when calling the set state methods so avoiding that if we can would be good?

Reply:
Yes, we should avoid setting objects that are different by reference but equal by value to prevent unnecessary rerendering.